### PR TITLE
Fix issues with `allow_nil: true`

### PR DESCRIPTION
**Pull request by joemsak**
Turns out, the documentation I added in #995 wasn't 100% accurate. I had
to add tests for such examples and improve the internal code to support
the option. This should now provide the fully expected behavior.

### CODE REVIEW

**Comment by joemsak** [comment]
@parndt I missed some necessary changes, this should resolve that.

File: lib/friendly_id/finder_methods.rb [hunk 1]
```diff
@@ -27,9 +27,20 @@ module FinderMethods
...
+      end
+
       first_by_friendly_id(id).tap { |result| return result unless result.nil? }
-      return super(*args) if potential_primary_key?(id)
+
+      begin
+        return super(*args) if potential_primary_key?(id)
+      rescue => e
+        return allow_nil ? nil : raise(e)
+      end
```
Comment (joemsak) [inline lib/friendly_id/finder_methods.rb:43]: I know we talked about exception control flow and I'm definitely not a fan, however I really am not aware of another strategy to use here. I'm open to suggestions though.

File: lib/friendly_id/finder_methods.rb [hunk 2]
```diff
@@ -31,7 +31,9 @@ def find(*args, allow_nil: false)
       first_by_friendly_id(id).tap { |result| return result unless result.nil? }
       return super(*args) if potential_primary_key?(id)
 
-      raise_not_found_exception(id) unless allow_nil
+      raise_not_found_exception(id)
+    rescue => e
+      allow_nil ? nil : raise(e)
```
Comment (joemsak) [inline lib/friendly_id/finder_methods.rb:36]: I know we talked about exception control flow and I'm definitely not a fan, however I really am not aware of another strategy to use here. I'm open to suggestions though.

**Comment by joemsak** [comment]
Dearest @parndt what is the likelihood of this getting merged and then a release? 🙏This PR satisfies the documentation. Let me know if you have feedback. Thanks!

File: lib/friendly_id/finder_methods.rb (hunk 2 above)
Comment (parndt) [inline lib/friendly_id/finder_methods.rb:36]:
```suggestion
      raise_not_found_exception(id) unless allow_nil
    rescue ActiveRecord::RecordNotFound => exception
      raise exception unless allow_nil
```

What about this version? If we avoid raising the exception in the first place then we get higher performance by avoiding a needless raise/rescue/raise cycle.

**Comment by parndt** [comment]
@joemsak I have been thinking about this change on and off all week! I've added a suggestion, let me know what you think

**Comment by parndt** [comment]
I also spotted a duplicate `raise` which I've removed

**Comment by joemsak** [comment]
Thank you!

**Review by parndt — APPROVED** [review]

**Comment by joemsak** [comment]
Hey @parndt any chance on this making it to a release? Let me know your thoughts, thanks!

**Comment by parndt** [comment]
@joemsak yes, definitely, I will get to it as soon as I can. 😄

**Comment by parndt** [comment]
Version 5.5.0 pushed @joemsak

**Pull request merged**